### DISCUSSION TITLE
Coordinate mapping

### DISCUSSION THREAD

**Discussion by dlrice** [molstar/molstar#110] [Q&A] [answered]
Firstly, thanks a lot for writing this excellent library. We are looking to replace the current structure visualisation tool in our codebase with Mol* and we're looking for some advice regarding coordinate mapping. For some context, a user provides a set of amino acid regions (eg `highlight="290:300,310:340`) of a protein which we then map to the coordinates for a PDB entry (using PDBe's https://www.ebi.ac.uk/pdbe/api/mappings/uniprot/:pdbid endpoint) which finally is highlighted in the Mol*'s structure viewer using a MolQL query. The question I have is: we are currently performing the coordinate mapping outside of Mol* but does this functionality exist within Mol*?

**Comment by dsehnal** [comment, marked as answer]
Glad to see new Mol* integration!

Can you please be more specific about how you are performing the mapping and how you are doing the highlight?
Then I would be better help you to potentially optimize this step.

**Reply by dlrice** [reply]
I've created a [notebook](https://github.com/dlrice/ebi-notebooks/blob/master/protein-to-structure-coordinate-mapping.ipynb) that walks through the process with an example.

**Reply by dsehnal** [reply]
I see. So you currently compute the highlight ranges on the server and you would like to do this on the client directly from the API?

**Reply by dlrice** [reply]
I can see why you would ask about the server given that I have shared a python notebook with you - sorry to add confusion to my question. We perform all of the coordinate mapping (fetching from PDBe, subtracting offset, etc) on the client and then this makes it way into Mol* via:

```
highlight(ranges: { start: number; end: number }[]): void {
    const data = this.plugin.managers.structure.hierarchy.current.structures[0]
      ?.cell.obj?.data;
    if (!data) return;
    const sel = Script.getStructureSelection(
      (Q) =>
        Q.struct.generator.atomGroups({
          "residue-test": Q.core.logic.or(
            ranges.map(({ start, end }) =>
              Q.core.rel.inRange([
                Q.struct.atomProperty.macromolecular.auth_seq_id(),
                start,
                end,
              ])
            )
          ),
        }),
      data
    );
    const loci = StructureSelection.toLociWithSourceUnits(sel);
    this.plugin.managers.interactivity.lociSelects.selectOnly({ loci });
  }
```

I was wondering if a MolQL query exists which would highlight/select the structure's coordinate range given a complete peptide chain (eg a UniProt isoform) coordinate range?

**Reply by dsehnal** [reply]
You could add your own symbol to the query language that does that and perhaps gain a marginal perf improvement which probably noone would notice.

In Mol* you have access to all categories of the input mmCIF file and you can make the transformations based on that if that's what you are asking. Otherwise if the coordinate system of the input file is different than what the API provides, you have to make the transformation by "hand".

Having said that, if it's a common transformation like UniProt -> PDB it might be worthwhile to add a utility function that helps with that.

**Reply by dlrice** [reply]
Will make the transformation by "hand" as you've mentioned for now. I guess time will tell how often UniProt -> PDB transformations will need to be made.
